sharding-orchestration/sharding-orchestration-config/sharding-orchestration-config-apollo/src/main/java/org/apache/shardingsphere/orchestration/config/apollo/ApolloConfigCenter.java

```diff
@@ -0,0 +1,130 @@
+import com.ctrip.framework.apollo.model.ConfigChangeEvent;
+import com.google.common.base.Strings;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.shardingsphere.orchestration.config.api.ConfigCenter;
+import org.apache.shardingsphere.orchestration.config.api.ConfigCenterConfiguration;
+import org.apache.shardingsphere.orchestration.config.listener.DataChangedEvent;
+import org.apache.shardingsphere.orchestration.config.listener.DataChangedEventListener;
+
+import java.util.List;
+import java.util.Properties;
+
+/**
+ * Registry center for Apollo.
+ *
+ * @author dongzonglei
+ */
+@Slf4j
+public final class ApolloConfigCenter implements ConfigCenter {
+
+    private Config apolloConfig;
+
+    @Getter
+    @Setter
+    private Properties properties = new Properties();
+
+    @Override
+    public void init(final ConfigCenterConfiguration config) {
+        System.getProperties().setProperty("app.id", properties.getProperty("appId", "APOLLO_SHARDING_SPHERE"));
+        System.getProperties().setProperty("env", properties.getProperty("env", "DEV"));
+        System.getProperties().setProperty(ConfigConsts.APOLLO_CLUSTER_KEY, properties.getProperty("clusterName", ConfigConsts.CLUSTER_NAME_DEFAULT));
+        System.getProperties().setProperty(ConfigConsts.APOLLO_META_KEY, config.getServerLists());
+        apolloConfig = ConfigService.getConfig(config.getNamespace());
+    }
+
+    @Override
+    public String get(final String key) {
+        return apolloConfig.getProperty(key.replace("/", "."), "");
+    }
+
+    @Override
+    public String getDirectly(final String key) {
+        return get(key);
+    }
+
+    @Override
+    public boolean isExisted(final String key) {
+        return !Strings.isNullOrEmpty(get(key));
+    }
+
+    @Override
+    public List<String> getChildrenKeys(final String key) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void persist(final String key, final String value) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void update(final String key, final String value) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void persistEphemeral(final String key, final String value) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void watch(final String key, final DataChangedEventListener dataChangedEventListener) {
+        apolloConfig.addChangeListener(new ConfigChangeListener() {
+            @Override
+            public void onChange(final ConfigChangeEvent changeEvent) {
+                for (String key : changeEvent.changedKeys()) {
+                    ConfigChange change = changeEvent.getChange(key);
+                    dataChangedEventListener.onChange(new DataChangedEvent(key, change.getNewValue(), getChangedType(change.getChangeType())));
+                }
+            }
+        });
```

Comment:
According to the method signature of `org.apache.shardingsphere.orchestration.config.api.ConfigCenter#watch`, (only watch a specific key's event) you should use `com.ctrip.framework.apollo.Config#addChangeListener(com.ctrip.framework.apollo.ConfigChangeListener, java.util.Set<java.lang.String>)` like following:

```java
        apolloConfig.addChangeListener(new ConfigChangeListener() {
            @Override
            public void onChange(final ConfigChangeEvent changeEvent) {
                for (String key : changeEvent.changedKeys()) {
                    ConfigChange change = changeEvent.getChange(key);
                    dataChangedEventListener.onChange(new DataChangedEvent(key, change.getNewValue(), getChangedType(change.getChangeType())));
                }
            }
        }, Sets.newHashSet(key));
```

Otherwise, the `dataChangedEventListener` will get notified whenever data of any key is updated, which may be inefficient.

Reply:
Hi @kezhenxu94 , Thank you very much. I have fixed the code like following:
```
apolloConfig.addChangeListener(new ConfigChangeListener() {
    @Override
    public void onChange(final ConfigChangeEvent changeEvent) {
        for (String key : changeEvent.changedKeys()) {
            ConfigChange change = changeEvent.getChange(key);
            DataChangedEvent.ChangedType changedType = getChangedType(change.getChangeType());
            if (DataChangedEvent.ChangedType.IGNORED != changedType) {
                dataChangedEventListener.onChange(new DataChangedEvent(key, change.getNewValue(), changedType));
            }
        }
    }
}, Sets.newHashSet(key));
```
I also add the `ChangedType` condition, if it is `IGNORED` status, I will do nothing.
If it's convenient for you to help me check it again.

Reply:
That looks good to me now